File: delivery_ops_bridge/services/scheduler.py
from __future__ import annotations

import threading
import time
from datetime import date, datetime
from typing import Callable, Dict

from ..models import utc_now_iso
from .dashboard import DashboardService
from .jobs import ScheduledJobs
from .message_handler import MessageHandler


JOB_TO_CONFIG_KEY = {
    "standup-push": "standup_push",
    "standup-second-remind": "standup_second_remind",
    "standup-mark-missing": "standup_mark_missing",
    "standup-summary": "standup_summary",
    "overdue-scan": "overdue_scan",
    "daily-summary": "daily_summary",
    "dashboard": "dashboard",
}


class InProcessScheduler:
    def __init__(
        self,
        get_handler: Callable[[], MessageHandler],
        interval_seconds: float = 30.0,
    ):
        self.get_handler = get_handler
        self.interval_seconds = interval_seconds
        self._stop = threading.Event()
        self._thread = threading.Thread(target=self._run, name="delivery-ops-scheduler", daemon=True)

    def start(self) -> None:
        if not self._thread.is_alive():
            self._thread.start()

    def stop(self) -> None:
        self._stop.set()
        self._thread.join(timeout=5)

    def _run(self) -> None:
        while not self._stop.is_set():
            try:
                self.tick(datetime.now())
            except Exception:
                try:
                    self.get_handler().store.append_audit_log("scheduler_error", {"at": utc_now_iso()})
                except Exception:
                    pass
            self._stop.wait(self.interval_seconds)

    def tick(self, now: datetime) -> Dict[str, str]:
        handler = self.get_handler()
        results: Dict[str, str] = {}
        for group in handler.config.groups:
            group_id = group.chat_id
            if not group_id:
                continue
            for job_name, config_key in JOB_TO_CONFIG_KEY.items():
                scheduled_time = group.schedule.get(config_key)
                if not self._is_due(scheduled_time, now):
                    continue
                if not group.schedule.get(f"{config_key}_enabled", True):
                    continue
                key = f"scheduled:{job_name}:{group_id}:{now.date().isoformat()}:{scheduled_time}"
                if handler.store.has_idempotency_key(key):
                    continue
                handler.store.set_idempotency_key(key, {"started_at": utc_now_iso(), "job_name": job_name, "group_id": group_id})
                try:
                    payload = self._run_job(handler, job_name, group_id, now.date())
                    handler.store.set_idempotency_key(
                        key,
                        {"completed_at": utc_now_iso(), "job_name": job_name, "group_id": group_id, "result": payload},
                    )
                    handler.store.append_audit_log(
                        "job_completed",
                        {"job_name": job_name, "group_id": group_id, "trigger": "scheduler", "result": payload},
                    )
                    results[f"{job_name}:{group_id}"] = "completed"
                except Exception as exc:
                    handler.store.set_idempotency_key(
                        key,
                        {"failed_at": utc_now_iso(), "job_name": job_name, "group_id": group_id, "error": str(exc)},
                    )
                    handler.store.append_audit_log(
                        "job_failed",
                        {"job_name": job_name, "group_id": group_id, "trigger": "scheduler", "error": str(exc)},
                    )
                    results[f"{job_name}:{group_id}"] = "failed"
        return results

    def _run_job(self, handler: MessageHandler, job_name: str, group_id: str, day: date):
        llm = handler.intent_parser.llm if handler.intent_parser else None
        dashboard = DashboardService(
            handler.store,
            handler.config.data_path,
            handler.config.project.name,
            handler.config.runtime.public_base_url,
        )
        jobs = ScheduledJobs(handler.config, handler.store, handler.feishu, dashboard, llm)
        methods = {
            "standup-push": jobs.standup_push,
            "standup-second-remind": jobs.standup_second_remind,
            "standup-mark-missing": jobs.standup_mark_missing,
            "standup-summary": jobs.standup_summary,
            "overdue-scan": jobs.overdue_scan,
            "daily-summary": jobs.daily_summary,
            "dashboard": jobs.dashboard_generate,
        }
        return methods[job_name](group_id, day)

    def _is_due(self, scheduled_time, now: datetime) -> bool:
        if not isinstance(scheduled_time, str):
            return False
        try:
            hour_text, minute_text = scheduled_time.split(":", 1)
            return int(hour_text) == now.hour and int(minute_text) == now.minute
        except (TypeError, ValueError):
            return False

File: backend/src/core/scheduler.ts
import { AppConfig } from './config';
import { PrismaStore } from './storage';
import { MessageHandler } from '@/services/messageHandler';
import { ScheduledJobs } from '@/services/jobs';
import { DashboardService } from '@/services/dashboard';
import { utcNowIso } from '@/models/types';
import dayjs from 'dayjs';

const JOB_TO_CONFIG_KEY: Record<string, string> = {
    "standup-push": "standup_push",
    "standup-second-remind": "standup_second_remind",
    "standup-mark-missing": "standup_mark_missing",
    "standup-summary": "standup_summary",
    "overdue-scan": "overdue_scan",
    "daily-summary": "daily_summary",
    "dashboard": "dashboard",
};

export class InProcessScheduler {
    private getHandler: () => MessageHandler;
    private intervalSeconds: number;
    private timer?: ReturnType<typeof setInterval>;

    constructor(getHandler: () => MessageHandler, intervalSeconds: number = 30) {
        this.getHandler = getHandler;
        this.intervalSeconds = intervalSeconds;
    }

    start() {
        if (this.timer) return;
        this.timer = setInterval(() => {
            try {
                this.tick(new Date());
            } catch (err) {
                try {
                    (this.getHandler() as any).ctx.store.appendAuditLog("scheduler_error", { at: utcNowIso() });
                } catch (e) {}
            }
        }, this.intervalSeconds * 1000);
    }

    stop() {
        if (this.timer) {
            clearInterval(this.timer);
            this.timer = undefined;
        }
    }

    async tick(now: Date): Promise<Record<string, string>> {
        const handler = this.getHandler();
        const ctx = (handler as any).ctx;
        const results: Record<string, string> = {};
        
        for (const group of ctx.config.groups) {
            const groupId = group.chat_id;
            if (!groupId) continue;

            for (const [jobName, configKey] of Object.entries(JOB_TO_CONFIG_KEY)) {
                const scheduledTime = (group.schedule as any)[configKey];
                if (!this.isDue(scheduledTime, now)) continue;
                
                if ((group.schedule as any)[`${configKey}_enabled`] === false) continue;
                
                const dateIso = dayjs(now).format('YYYY-MM-DD');
                const key = `scheduled:${jobName}:${groupId}:${dateIso}:${scheduledTime}`;
                
                if (await ctx.store.hasIdempotencyKey(key)) continue;
                
                await ctx.store.setIdempotencyKey(key, { started_at: utcNowIso(), job_name: jobName, group_id: groupId });
                
                try {
                    const payload = await this.runJob(handler, jobName, groupId, now);
                    await ctx.store.setIdempotencyKey(key, { completed_at: utcNowIso(), job_name: jobName, group_id: groupId, result: payload });
                    await ctx.store.appendAuditLog("job_completed", { job_name: jobName, group_id: groupId, trigger: "scheduler", result: payload });
                    results[`${jobName}:${groupId}`] = "completed";
                } catch (exc: any) {
                    await ctx.store.setIdempotencyKey(key, { failed_at: utcNowIso(), job_name: jobName, group_id: groupId, error: exc.message || String(exc) });
                    await ctx.store.appendAuditLog("job_failed", { job_name: jobName, group_id: groupId, trigger: "scheduler", error: exc.message || String(exc) });
                    results[`${jobName}:${groupId}`] = "failed";
                }
            }
        }
        return results;
    }

    private async runJob(handler: MessageHandler, jobName: string, groupId: string, day: Date) {
        const ctx = (handler as any).ctx;
        const llm = ctx.intentParser?.llm;
        const dashboard = new DashboardService(
            ctx.store,
            ctx.config.data_path,
            ctx.config.project.name,
            ctx.config.runtime.public_base_url
        );
        const jobs = new ScheduledJobs(ctx.config, ctx.store, ctx.feishu, dashboard, ctx.tapd, llm);
        
        switch (jobName) {
            case "standup-push": return await jobs.standupPush(groupId, day);
            case "standup-second-remind": return await jobs.standupSecondRemind(groupId, day);
            case "standup-mark-missing": return await jobs.standupMarkMissing(groupId, day);
            case "standup-summary": return await jobs.standupSummary(groupId, day);
            case "overdue-scan": return await jobs.overdueScan(groupId, day);
            case "daily-summary": return await jobs.dailySummary(groupId, day);
            case "dashboard": return await jobs.dashboardGenerate(groupId, day);
            default: throw new Error(`Unknown job: ${jobName}`);
        }
    }

    private isDue(scheduledTime: any, now: Date): boolean {
        if (typeof scheduledTime !== 'string') return false;
        try {
            const [hourText, minuteText] = scheduledTime.split(":", 2);
            return parseInt(hourText, 10) === now.getHours() && parseInt(minuteText, 10) === now.getMinutes();
        } catch {
            return false;
        }
    }
}
